### DISCUSSION TITLE
APIGW Response using Lambda custom (non-proxy) integration

### DISCUSSION THREAD

**Discussion by ymwjbxxq** [aws/aws-lambda-rust-runtime#548] [Q&A]
Hello all,

we have 2 libs:
- lambda_http
- lambda_runtime

And you cannot mix them, and I have a third case, which I can only solve with lambda_runtime trying to return a lambda_http response.

**PROBLEM:**
My APIGW has a Lambda Request Authorizer. The output of the Authorizer is a context, for example:
`
context: {
   email: a@a.com
}
`

To pass the context to Lambda, you must disable the LambdaProxy and create an integration template.
Using the default one from AWS, I end up having a payload that is not compatible with the types of lambda_http
```
##  See http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
##  This template will pass through all parameters including path, querystring, header, stage variables, and context through to the integration endpoint via the body/payload
#set($allParams = $input.params())
{
"body" : $input.json('$'),
"params" : {
#foreach($type in $allParams.keySet())
    #set($params = $allParams.get($type))
"$type" : {
    #foreach($paramName in $params.keySet())
    "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
        #if($foreach.hasNext),#end
    #end
}
    #if($foreach.hasNext),#end
#end
},
"stage-variables" : {
#foreach($key in $stageVariables.keySet())
"$key" : "$util.escapeJavaScript($stageVariables.get($key))"
    #if($foreach.hasNext),#end
#end
},
"auth" : {
    "email" : "$context.authorizer.email"
    }
}
```

Because of this, I need to create my custom structure and switch to lambda_runtime.

**LINKS:**
- https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-output.html
- https://aws.amazon.com/premiumsupport/knowledge-center/custom-headers-api-gateway-lambda/#:~:text=To%20pass%20custom%20headers%20from,from%20the%20original%20API%20request

**using lambda_http:**
```
 lambda_http::run(service_fn(|event: Request| {
        execute(event)
    }))
    .await?;
...

pub async fn execute(event: Request) -> Result<impl IntoResponse, Error> {
     let request = MyStruct::deserialize(MapDeserializer::new(result.body.into_iter())).unwrap();
    ...my code...
    
    Ok(response)
```

**I get this error:**
**ERROR lambda_runtime: Error("data did not match any variant of untagged enum LambdaRequest", line: 0, column: 0)**

**using lambda_runtime:**
```
#[derive(Debug, Serialize, Deserialize)]
pub struct Payload {
    #[serde(rename = "auth")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth: Option<Auth>,

    #[serde(rename = "body")]
    pub body: HashMap<String, Value>,

    #[serde(rename = "params")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Params>,
}

 lambda_runtime::run(service_fn(|event: LambdaEvent<Value>| {
        execute(event)
    }))
    .await?;

pub async fn execute(event: LambdaEvent<Value>) -> Result<Value, Error> {
     let request = MyStruct::deserialize(MapDeserializer::new(result.body.into_iter())).unwrap();
    ...my code...
    
    Ok(response)
```

**The problem is the response.** End of the day, I need to return something like the:
```
{
  "statusCode": 200,
  "body": "ok",
  "headers": { 
    "Access-Control-Allow-Origin": "*",
    ...
    "Accept": "application/json"
  }
}
```

If I use:
serde_json::from_str(response)?

It will return all as JSON in the body, and so it will skip the headers (see images)

![body](https://user-images.githubusercontent.com/78874812/195395284-0b26d4ad-a7a3-4cde-9faf-688dd408199a.png)
![missing_headers](https://user-images.githubusercontent.com/78874812/195395288-26a24b43-e0e8-43ca-a8d6-b026c763592f.png)

### **Question is:**
- how can I do it?

If I return `Result<Response<String>, Error` lambda_runtime complain

```
 the trait bound `lambda_http::Response<std::string::String>: Serialize` is not satisfied
the following other types implement trait `Serialize`:
```

**Comment by calavera** [comment]
`lambda_http` is an abstraction that takes payloads from different services and turns them into http objects. If you remove the abstraction, you need to return something that the service that you're using expects.

If you're using APIGW Rest, you probably need to return a ProxyResponse object:

https://github.com/LegNeato/aws-lambda-events/blob/master/aws_lambda_events/src/apigw/mod.rs#L60

If you're using APIGW Http, you probably need to return a HttpResponse object:

https://github.com/LegNeato/aws-lambda-events/blob/master/aws_lambda_events/src/apigw/mod.rs#L295

**Reply by ymwjbxxq** [reply]
@calavera 

I think there is something else.

I set the response in this way:
```
#[tokio::main]
async fn main() -> Result<(), Error> {
    lambda_runtime::run(service_fn(|event: LambdaEvent<Value>| {
        execute(event)
    }).await?;
    Ok(())
}

pub async fn execute(event: LambdaEvent<Value>) -> Result<ApiGatewayProxyResponse, Error> {
  let mut headers = HeaderMap::new();
  headers.insert(http::header::CONTENT_TYPE, HeaderValue::from_static("application/json"));
  headers.insert(http::header::ACCESS_CONTROL_ALLOW_ORIGIN, HeaderValue::from_static("*"));
  headers.insert(http::header::ACCESS_CONTROL_ALLOW_HEADERS, HeaderValue::from_static("Content-Type"));
  headers.insert(http::header::ACCESS_CONTROL_ALLOW_METHODS, HeaderValue::from_static("GET, POST, OPTIONS, PATCH, PUT, DELETE"));
  headers.insert(http::header::ACCESS_CONTROL_ALLOW_CREDENTIALS, HeaderValue::from_static("true"));
  
   let response = ApiGatewayProxyResponse {
      status_code: 200,
      headers,
      multi_value_headers: HeaderMap::new(),
      body: Some(body),
      is_base64_encoded: None,
  };
  
  Ok(response)
}

```

**but it still returning all in the body of the response like in the images**
![body](https://user-images.githubusercontent.com/78874812/195395284-0b26d4ad-a7a3-4cde-9faf-688dd408199a.png)
![missing_headers](https://user-images.githubusercontent.com/78874812/195395288-26a24b43-e0e8-43ca-a8d6-b026c763592f.png)


Whether with TypeScript, I use types or not (any), it works until the return is compatible.

Is it possible that the lambda_runtime response is not compatible with APIGW?

**Reply by calavera** [reply]
> Is it possible that the lambda_runtime response is not compatible with APIGW?

What needs to be compatible is the structure that's serialized as JSON for APIGW to read. What your code is returning is compatible with APIGW. For your screenshot, it looks like your APIGW configuration might be the problem in this case. I don't know if the Authorizer is involved in this.

**Reply by ymwjbxxq** [reply]
at this point is just APIGW and Lambda
I will investigate the integration response template of the APIGW

The default is (not working):
```
##  See http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html
##  This template will pass through all parameters including path, querystring, header, stage variables, and context through to the integration endpoint via the body/payload
#set($allParams = $input.params())
{
"body-json" : $input.json('$'),
"params" : {
#foreach($type in $allParams.keySet())
    #set($params = $allParams.get($type))
"$type" : {
    #foreach($paramName in $params.keySet())
    "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
        #if($foreach.hasNext),#end
    #end
}
    #if($foreach.hasNext),#end
#end
},
"stage-variables" : {
#foreach($key in $stageVariables.keySet())
"$key" : "$util.escapeJavaScript($stageVariables.get($key))"
    #if($foreach.hasNext),#end
#end
},
"context" : {
    "account-id" : "$context.identity.accountId",
    "api-id" : "$context.apiId",
    "api-key" : "$context.identity.apiKey",
    "authorizer-principal-id" : "$context.authorizer.principalId",
    "caller" : "$context.identity.caller",
    "cognito-authentication-provider" : "$context.identity.cognitoAuthenticationProvider",
    "cognito-authentication-type" : "$context.identity.cognitoAuthenticationType",
    "cognito-identity-id" : "$context.identity.cognitoIdentityId",
    "cognito-identity-pool-id" : "$context.identity.cognitoIdentityPoolId",
    "http-method" : "$context.httpMethod",
    "stage" : "$context.stage",
    "source-ip" : "$context.identity.sourceIp",
    "user" : "$context.identity.user",
    "user-agent" : "$context.identity.userAgent",
    "user-arn" : "$context.identity.userArn",
    "request-id" : "$context.requestId",
    "resource-id" : "$context.resourceId",
    "resource-path" : "$context.resourcePath"
    }
}

```

I keep you posted

**Reply by calavera** [reply]
I wonder if the problem is this statement `"body-json" : $input.json('$')`. Again, I have no idea how that works 😅 but the way I'd read that is "send all the information in th body-json".

**Reply by bnusunny** [reply]
APIGW integration response templates apply to the responses from backend integration endpoints, not the responses from Lambda Authorizers. 

The output from Lambda Authorizer needs to be in a specific format, as documented here: 
REST API: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-output.html 
HTTP API: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-lambda-authorizer.html

I don't think lambda-http crate will work for authorizer. Lambda-runtime crate is the way to implement it.

**Reply by ymwjbxxq** [reply]
Let me sum up:
- lambda_http does not work at the time you want to pass the output of the Lambda Authorizer to the Lambda behind the APIGW (non-proxy settings)
- to handle this case, you need to use lambda_runtime (create a custom integration request template), and the input of the handler is a custom-made structure
- the Response of the Lambda can be ApiGatewayProxyResponse, but APIGW will not return the correct settings out of the box
- I must create a custom integration response template to map the headers correctly and the body (working in progress)

**Reply by bnusunny** [reply]
The last two steps can be simplified. With lambda-runtime, you can directly return the data in the required format. The basic lambda example gives a good example [here](https://github.com/awslabs/aws-lambda-rust-runtime/blob/a0f6828ccf6a42643eef91ef48f38f0dada52218/examples/basic-lambda/src/main.rs#L43-L46).

**Reply by bnusunny** [reply]
Are you having issues with the response for Lambda Authorizer or that for a Lambda function as an integration target?

**Reply by ymwjbxxq** [reply]
The Lambda Authorizer is fine to return the correct type (ApiGatewayCustomAuthorizerResponse). 
The APIGW pass the output of the Lambda Authorizer to the Lambda and I can read it.
![Screenshot 2022-10-14 at 09 34 37](https://user-images.githubusercontent.com/78874812/195789279-fb1c7c47-a08a-4934-8315-848372edec20.png)

because I created a mapping template:
![Screenshot 2022-10-14 at 09 35 23](https://user-images.githubusercontent.com/78874812/195789511-0e220b29-e4fd-4102-bc1d-d2a9be92a933.png)


I am having issues with the integration response
![Screenshot 2022-10-14 at 09 32 09](https://user-images.githubusercontent.com/78874812/195788797-aa8478ee-2f39-4299-abaa-a81c5793892b.png)

trying mapping the Header Mappings and status from the Lambda response (ApiGatewayProxyResponse)

**Reply by bnusunny** [reply]
Okay. Got it. Creating a generic response mapping from lambda proxy response to APIGW response sounds like a good idea.

**Reply by ymwjbxxq** [reply]
I made it with lambda_runtime

```
 lambda_runtime::run(service_fn(|event: LambdaEvent<Value>| {
        execute(event)
    }))
    .await?;

pub async fn execute(event: LambdaEvent<Value>) -> Result<Value, MyCustomError> {
     let result: Payload = serde_json::from_value(event.payload).unwrap();
     let request = MyStruct::deserialize(MapDeserializer::new(result.body.into_iter())).unwrap();
    //handler logic

    //return 200
    return Ok(json!({ "message": "all good" }));
    
    // imagine 404
    Err(ApplicationError::InternalError(format!(
        "errors: {:?}",
        errors.join("##")
    )))
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Payload {
    #[serde(rename = "auth")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth: Option<Auth>,

    #[serde(rename = "body")]
    pub body: HashMap<String, Value>,

    #[serde(rename = "params")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Params>,
}
```

Now we must have 2 mapping templates in the APIGW:
-  integration request
- integration response

**Integration request:**
```
#set($allParams = $input.params())
"body": $input.json('$'),
"params" : {
#foreach($type in $allParams.keySet())
   #set($params = $allParams.get($type))
   "$type" : {
        #foreach($paramName in $params.keySet())
              "$paramName" : "$util.escapeJavaScript($params.get($paramName))"
               #if($foreach.hasNext),#end
       #end 
    }
    #if($foreach.hasNext),#end
#end 
},
"stage-variables" : {
#foreach($key in $stageVariables.keySet())
  "$key" : "$util.escapeJavaScript($stageVariables.get($key))"
    #if($foreach.hasNext),#end
#end},
"auth" : {
    "myprop":"$context.authorizer.myprop" ##custom output from the Lambda Authorizer
}
}"
```

**Integration response:**

- Here you need to map all the status code that the Lambda return, for example, 200 and 404
- For errors, you need to match the string of the error message, so I solve putting a text specifically for the error code that I will remove in the mapping. And with replacing and split I will construct the output JSON that I want.
Maybe there is a better way to do it but I could not find it. At least you have the idea ;)

```
#set( $response = $input.path('$.errorMessage').replace("errors: ", ""))
#set( $errors = $response.replace('\"', ""))
{ 
"errors": [
#foreach($item in $errors.split('##'))
       "$item"
      #if($foreach.hasNext),#end
#end  
]
}
```

**Reply by calavera** [reply]
that's awesome! @ymwjbxxq what do you think about creating an example in the examples directory, I'm sure other people would appreciate it, it they might not be able to find it here.
